### DISCUSSION TITLE
Is there a camera limit smoothing alternative on PhantomCamera2D ?

### DISCUSSION THREAD

**Discussion by ClementMONDARY** [ramokz/phantom-camera#411] [Question / Help Wanted] [answered]
Im currently working on a scene with a first PhantomCamera2D following the player, and a second for the main menu. They are both glued to the player and got limits set to the same values. 'Follow Dampin' is on, the cameras work smoothly when moving.

The 2 problems im currently having is when approching the cameras limits :

## 1. No limit smoothing :
I did not found any options similar to 'Limit > Smoothed' option of Camera2D in PhantomCamera2D. The camera movement just stops instantly when it hits the limits.

[No limit smoothing.webm](https://github.com/user-attachments/assets/da91b5f8-b8fe-4a80-b65f-8ca403c6d6a0)

## 2. Forget the limits when swaping between cams :
When opening or closing the menu just next to the borders, wich are using different PhantomCamera2D, they do not take in account the fixed limit borders, resulting in an instant correction when finishing the animation.

[Forget the limits.webm](https://github.com/user-attachments/assets/f5f90572-658b-4a15-aada-305e14d5f656)

## Here are the scene's tree structure and camera settings for better understanding :

![image](https://github.com/user-attachments/assets/83eddb10-9fbb-4a43-b380-1d0e08c092f6)

![image](https://github.com/user-attachments/assets/00df73ae-6ab4-4cc9-94d2-03e62952930f)

As you can imagine, these questions are not blocking anything, but if there are alternatives to solve them, whether negative or positive, I'd be delighted. If there's no solution, maybe these could be potential features to add ? (in the limits of possible)

**Comment by ramokz** [comment]
> I did not found any options similar to 'Limit > Smoothed' option of Camera2D in PhantomCamera2D. The camera movement just stops instantly when it hits the limits.

This was something I did briefly explore when I initially added the `Limit` support for the addon. I didn't get around to check how Godot does the `Smoothed` behavior internally, but it's on my radar to create better feature parity.

> When opening or closing the menu just next to the borders, wich are using different PhantomCamera2D, they do not take in account the fixed limit borders, resulting in an instant correction when finishing the animation.

That one is odd, the `PCam2D` nodes should be confined to the limit at all times, even when inactive — can see an example in the `2d_limit_example_scene.tscn` inside `examples/example_scenes`. Are you changing the `limit` value at any point before, during, or after the transition?

**Reply by ClementMONDARY** [reply]
> That one is odd, the `PCam2D` nodes should be confined to the limit at all times, even when inactive — can see an example in the `2d_limit_example_scene.tscn` inside `examples/example_scenes`. Are you changing the `limit` value at any point before, during, or after the transition?

The `Limit` value is not touch at any point, on both cams. I even tried to set the `Limit Target` assignation to a `CollisionShape2D`, with no success.

## Issues only happening in fullscreen

I was doing some tests earlier and noticed that the issues only happen in FULLSCREEN.
When on default window parameters (1152x648), the `Follow Damping` when approching borders and the `PCam2D` changing priorities when opening the menu work perfectly. The same issues are happening in the `2d_limit_example_scene.tscn` that you mentioned earlier.

Exemple on my scene :

[20e41dec-3125-46bb-bd4b-293caf607c9c.webm](https://github.com/user-attachments/assets/fa81ac6a-a5fc-4a2f-b413-9a35ed25b630)

Exemple on `2d_limit_example_scene.tscn` scene :

[da370bce-806c-4572-9879-536a4211733b.webm](https://github.com/user-attachments/assets/3c5aa6a2-6c85-4159-a045-0163a5da1a40)

Could these be fixed by changing the projects settings (that i have not touched for this project), a cam setting that i have skipped, or have they just not been foreseen ?

**Reply by ramokz** [reply, marked as answer]
Think that happens because you're using the default (`disabled`) `Stretch Mode` inside Project Settings?
As it would mean that as the window gets bigger, so too does the Viewport Width / Height of the camera. The complication is that if you have a relatively compact limit region and increase the window size too much, then the camera's viewport may extend beyond the bounds of the limit. Which can result in the PCam2D getting confused as it's now bigger than the area it's confined too.

If you change the Viewport Width / Height inside Project Settings to match your full-screen resolution, and see how it behaves in the editor, then might give a better clue for how it behaves during runtime.

Happy to take a look if you're willing to share a sample of the project scene in question.

**Reply by ramokz** [reply]
It does feel like there's a way to prevent the `PCam2D` from freaking out the way it currently does, but I am not too sure what that solution looks like.

**Reply by ClementMONDARY** [reply]
Modifying the strech mode in `Display > Window > Stretch > Mode` works fine for me.
I went to the wiki to see what those parameters mean, and that's exactly what i wanted. It makes sense, since the PCam2d limits are designed to work with the resolution of the project. Making the window stretch prevents this from happening, as well as being useful for future in-game options.

Thank you for solving my problem @ramokz. if you want to take a look at my project, you can found it here in -> [https://github.com/ClementMONDARY/Tree-and-Inventory](url)

**Reply by ramokz** [reply]
No problem!

There are still use cases where you would want to scale the camera based on the monitor size if you want to avoid letterboxing, though that's a game design / art directional choice more than anything else.

Think there is still room to make the limit mechanic better handle increased resolutions down the line.
